### PR TITLE
test: run e2e on isloated directory

### PR DESCRIPTION
**Pull request by hung-cybo**
## Why

We want to run cli-kintone on an isolated directory during E2E testing.

Therefore, the created input/output files during E2E testing do not affect the test result.

## What
- [x] Running cli-kintone on a temporal directory during E2E testing.
- [x] Removing the temporal directory if there is no failed scenario.

## How to test

```
pnpm build
pnpm test:e2e
```

## Checklist

- [x] Read [CONTRIBUTING.md](https://github.com/kintone/cli-kintone/blob/main/CONTRIBUTING.md)
- [x] Updated documentation if it is required.
- [x] Added/updated tests if it is required. (or tested manually)
- [x] Passed `pnpm lint` and `pnpm test` on the root directory.

### CODE REVIEW

File: cucumber.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ const config = {
   default: {
     formatOptions: { snippetInterface: "synchronous" },
     requireModule: ["ts-node/register"],
-    require: ["features/step_definitions/**/*.ts"],
+    require: ["features/support/**/*.ts", "features/step_definitions/**/*.ts"],
```
Comment (tasshi-me) [inline cucumber.js]: NITS: Please rename `support` to `supports`.

File: features/ultils/helper.ts [hunk 2]
```diff
@@ -45,6 +49,7 @@ const replacer = (substring: string) => {
 
 export const createCsvFile = async (
   inputCsvObject: string[][],
+  workingDir?: string,
   destFilePath?: string,
```
Comment (tasshi-me) [inline features/ultils/helper.ts]: Please wrap these two inputs into `options` because both of them are `string` and can make mistakes.

File: features/ultils/world.ts [hunk 3]
```diff
@@ -17,6 +19,18 @@ export class OurWorld extends World {
...
+    this.workingDir = options.rootDir;
+    this.log(`Root working directory: ${this.workingDir}`);
+  }
+
+  public initForIsolatedScenario() {
+    this.workingDir = fs.mkdtempSync(
+      this.workingDir ? path.join(this.workingDir, "case-") : "case-",
+    );
+    this.log(`Scenario working directory: ${this.workingDir}`);
+  }
```
Comment (tasshi-me) [inline features/ultils/world.ts]: I think the `init()` should know only the `workingDir` where the World should work at.

I mean, the processes in `initForIsolatedScenario` should be performed in `Before({ tags: "@isolated" }`

File: features/ultils/helper.ts [hunk 4]
```diff
@@ -45,6 +49,7 @@ const replacer = (substring: string) => {
 
 export const createCsvFile = async (
   inputCsvObject: string[][],
+  workingDir?: string,
```
Comment (tasshi-me) [inline features/ultils/helper.ts]: In the `createCsvFile`, how about naming `baseDir` than `workingDir` so that we can know how that option will be used in the function?

File: cucumber.js (hunk 1 above)
Comment (tasshi-me) [reply cucumber.js]: NOTE: Does the name `supports` make it understandable that the files inside set up some meta configurations?

=> Maybe yes. In the official document, the parent directory of the `World` and `Hooks` is `support_files`.
https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/world.md

File: features/support/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+Before({ tags: "@isolated" }, function () {
+  this.initForIsolatedScenario();
+});
+
+After(async function (scenario) {
+  if (scenario.result?.status === Status.FAILED) {
+    failedScenarioCount++;
+  }
+});
```
Comment (tasshi-me) [inline features/support/hooks.ts]: The `Before` and `After` accept `World` as their generics parameters.
Please define helper functions on `world.ts` as the same as `Given`/`When`/`Then`.

NOTE: With default `Before` and `After`, the type of `this[parameter]` is any.

File: features/support/hooks.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  Before,
+  BeforeAll,
+  setWorldConstructor,
+  Status,
+} from "@cucumber/cucumber";
+import fs from "fs";
+import path from "path";
+import os from "os";
+
+setWorldConstructor(OurWorld);
```
Comment (tasshi-me) [inline features/support/hooks.ts]: Do we need `setWorldConstructor()` here?

**Review by tasshi-me — CHANGES_REQUESTED** [review]
Thank you for your PR!
I left some reviews.

File: cucumber.js (hunk 1 above)
Comment (hung-cybo) [reply cucumber.js]: The name `support` is from the official document, [ref](https://cucumber.io/docs/cucumber/api/?sbsearch=hook&lang=javascript#hooks).

But, I think `supports` is more understandable. And the same style as other folders, "step_definitions", "features"
I will update it.

File: features/ultils/world.ts (hunk 3 above)
Comment (hung-cybo) [reply features/ultils/world.ts]: @mshrtsr 

Could you explain more details about this?
Now, we have two Before hooks.

```
// Because World isn't available in BeforeAll, we have to use Before hook.
Before(function () {
  this.init({ rootDir });
});

Before({ tags: "@isolated" }, function () {
  this.initForIsolatedScenario();
});
```

File: features/ultils/helper.ts (hunk 4 above)
Comment (hung-cybo) [reply features/ultils/helper.ts]: Oh, you're right.
It should be `baseDir`, not `workingDir`.
Thank you.

File: features/support/hooks.ts (hunk 6 above)
Comment (hung-cybo) [reply features/support/hooks.ts]: It's redundant. I will remove it.

File: cucumber.js (hunk 1 above)
Comment (tasshi-me) [reply cucumber.js]: @hung-cybo 
Oh, I didn't check that document!
I wondered if `support/supports` are not understandable as other ideas: `configs/setups/environments`

Yes, please go with `supports`.

File: features/ultils/world.ts (hunk 3 above)
Comment (mshrtsr) [reply features/ultils/world.ts:33]: I mean the following.
Only the `Before()` hook know about the `isolated`.

## In hooks.ts

```typescript
BeforeAll(function () {
  rootDir = fs.mkdtempSync(
    path.join(os.tmpdir(), `cli-kintone-e2e-test-${new Date().valueOf()}-`),
  );
  console.log(`Root working directory: ${rootDir}`)
});

Before(function () {
  this.init({ workingDir: rootDir });
});

Before({ tags: "@isolated" }, function () {
  const workingDir = fs.mkdtempSync(path.join(rootDir, "case-") : "case-");
  this.init({ workingDir });
});
```

## In world.ts

```typescript
export class OurWorld extends World {
  public workingDir?: string;

  ...

  public init(options: { workingDir: string }) {
    this.workingDir = options.workingDir;
    this.log(`Scenario working directory: ${this.workingDir}`);
  }

  // Remove initForIsolatedScenario()
}
```

Comment (tasshi-me) [reply features/ultils/world.ts]: @hung-cybo 
I mean the following.
Only the `Before()` hook knows about the `isolated`.

## In hooks.ts

```typescript
BeforeAll(function () {
  rootDir = fs.mkdtempSync(
    path.join(os.tmpdir(), `cli-kintone-e2e-test-${new Date().valueOf()}-`),
  );
  console.log(`Root working directory: ${rootDir}`)
});

Before(function () {
  this.init({ workingDir: rootDir });
});

Before({ tags: "@isolated" }, function () {
  const workingDir = fs.mkdtempSync(path.join(rootDir, "case-") : "case-");
  this.init({ workingDir });
});
```

## In world.ts

```typescript
export class OurWorld extends World {
  public workingDir?: string;

  ...

  public init(options: { workingDir: string }) {
    this.workingDir = options.workingDir;
    this.log(`Scenario working directory: ${this.workingDir}`);
  }

  // Remove initForIsolatedScenario()
}
```

File: features/supports/world.ts [hunk 7]
```diff
@@ -1,10 +1,12 @@
+import type { SpawnSyncReturns } from "child_process";
```
Comment (tasshi-me) [inline features/supports/world.ts]: Oh, I think we don't need to put this file in the `supports` directory.
The files in the `supports` directory are loaded automatically.
So that we can use Hooks and `setWorldConstructor()` in support codes(`hooks.ts` and `setup.ts`)

On the other hand, This file doesn't have actual function calls.
This file `world.ts` has only a definition of the `OurWorld` class and some wrapper functions.
We use them on the support codes by importing.

In the document, they put their World definition outside (`../classes/CustomWorld.js`) and import it in support code (`setup.js`)
https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/world.md#real-world-example

In another idea, if we put put `world.ts` to `supports`, we can call `setWorldConstructor()` in `world.ts` and delete `setup.ts`.

In summary, it's better to choose one of the following:
- Move back `world.ts` to `utls`
- Call `setWorldConstructor()` in `world.ts` and delete `setup.ts`

File: features/supports/world.ts [hunk 8]
```diff
@@ -17,9 +18,29 @@ export class OurWorld extends World {
...
   }
+
+  public init(options: { workingDir: string }) {
+    this.workingDir = options.workingDir;
+  }
+
+  public execCliKintoneSync(args: string) {
+    this.response = execCliKintoneSync(args, {
+      env: this.env,
+      cwd: this.workingDir,
```
Comment (tasshi-me) [inline features/supports/world.ts]: NOTE:
It's more secure to implement a getter the same as `this.response` to prevent unexpected behavior when `this.workingDir` is `undefined`.
It's nice to have, but we don't need it now.

**Review by tasshi-me — CHANGES_REQUESTED** [review]
@hung-cybo 
I have one question about the location of `world.ts`.
Others are LGTM.

File: features/supports/world.ts (hunk 7 above)
Comment (hung-cybo) [reply features/supports/world.ts]: @mshrtsr 

OIC. Let's follow the document.
I moved `world.ts` back to `utils` folder.
Thank you.

**Review by tasshi-me — APPROVED** [review]
LGTM!!!
